### PR TITLE
Add support for timestamp with local zone in Hive3

### PR DESCRIPTION
**Pull request by marton-bod**
- Break up `TimestampObjectInspector` classes into two separate classes for (i) normal timestamps and (ii) timestamps with zone (instead of using inner classes)
- For Hive3, take advantage of its TimestampWithLocalTimeZone type (which Hive2 doesn't have) instead of just converting timestampTz into normal timestamps

### CODE REVIEW

**Comment by marton-bod** [comment]
@pvary can you please review when you have the chance?

File: hive3/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspectorHive3.java [hunk 1]
```diff
@@ -20,50 +20,32 @@
...
-import java.time.OffsetDateTime;
 import java.time.ZoneOffset;
 import org.apache.hadoop.hive.common.type.Timestamp;
 import org.apache.hadoop.hive.serde2.io.TimestampWritableV2;
 import org.apache.hadoop.hive.serde2.objectinspector.primitive.AbstractPrimitiveJavaObjectInspector;
 import org.apache.hadoop.hive.serde2.objectinspector.primitive.TimestampObjectInspector;
 import org.apache.hadoop.hive.serde2.typeinfo.TypeInfoFactory;
 
-public abstract class IcebergTimestampObjectInspectorHive3 extends AbstractPrimitiveJavaObjectInspector
+public class IcebergTimestampObjectInspectorHive3 extends AbstractPrimitiveJavaObjectInspector
```
Comment (pvary) [inline hive3/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspectorHive3.java:30]: You might want to rebase to master. We need to implement WriteObjectInspector.
Also we need to test for them as well

File: hive3/src/test/java/org/apache/iceberg/mr/hive/serde/objectinspector/TestIcebergTimestampWithZoneObjectInspectorHive3.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    Assert.assertEquals(TimestampTZ.class, oi.getJavaPrimitiveClass());
+    Assert.assertEquals(TimestampLocalTZWritable.class, oi.getPrimitiveWritableClass());
+
+    Assert.assertNull(oi.copyObject(null));
+    Assert.assertNull(oi.getPrimitiveJavaObject(null));
+    Assert.assertNull(oi.getPrimitiveWritableObject(null));
+
+    long epochSeconds = 1601471970L;
+    OffsetDateTime offsetDateTime = OffsetDateTime.of(
```
Comment (pvary) [inline hive3/src/test/java/org/apache/iceberg/mr/hive/serde/objectinspector/TestIcebergTimestampWithZoneObjectInspectorHive3.java:54]: Could we test that the same instant with different offset will result in the same TimestampTZ?

File: hive3/src/test/java/org/apache/iceberg/mr/hive/serde/objectinspector/TestIcebergTimestampWithZoneObjectInspectorHive3.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    Assert.assertEquals(TypeInfoFactory.timestampLocalTZTypeInfo.getTypeName(), oi.getTypeName());
+
+    Assert.assertEquals(TimestampTZ.class, oi.getJavaPrimitiveClass());
+    Assert.assertEquals(TimestampLocalTZWritable.class, oi.getPrimitiveWritableClass());
+
+    Assert.assertNull(oi.copyObject(null));
+    Assert.assertNull(oi.getPrimitiveJavaObject(null));
+    Assert.assertNull(oi.getPrimitiveWritableObject(null));
+
+    long epochSeconds = 1601471970L;
```
Comment (pvary) [inline hive3/src/test/java/org/apache/iceberg/mr/hive/serde/objectinspector/TestIcebergTimestampWithZoneObjectInspectorHive3.java:53]: How can we be sure that this epoch seconds really result in the expected TZ?

File: mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspector.java [hunk 4]
```diff
@@ -80,22 +57,16 @@ public TimestampWritable getPrimitiveWritableObject(Object o) {
 
   @Override
   public Object copyObject(Object o) {
-    if (o == null) {
```
Comment (pvary) [inline mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspector.java:83]: Why did we remove this?

File: hive3/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspectorHive3.java (hunk 1 above)
Comment (marton-bod) [reply hive3/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspectorHive3.java:30]: I've done a merge with master, and this class `IcebergTimestampObjectInspectorHive3` doesn't seem to implement `WriteObjectInspector` yet even on master. I can add a similar convert method that we have for `IcebergTimestampObjectInspector`.

File: hive3/src/test/java/org/apache/iceberg/mr/hive/serde/objectinspector/TestIcebergTimestampWithZoneObjectInspectorHive3.java (hunk 2 above)
Comment (marton-bod) [reply hive3/src/test/java/org/apache/iceberg/mr/hive/serde/objectinspector/TestIcebergTimestampWithZoneObjectInspectorHive3.java:54]: Yes, that's a good idea.

**Comment by pvary** [comment]
Do we need to change `HiveSchemaConverter`?

File: hive3/src/test/java/org/apache/iceberg/mr/hive/serde/objectinspector/TestIcebergTimestampWithZoneObjectInspectorHive3.java (hunk 3 above)
Comment (marton-bod) [reply hive3/src/test/java/org/apache/iceberg/mr/hive/serde/objectinspector/TestIcebergTimestampWithZoneObjectInspectorHive3.java:53]: Refactored this to use LocalDateTime instead

File: mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspector.java (hunk 4 above)
Comment (marton-bod) [reply mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspector.java:83]: this null check is handled in the `else` branch

**Comment by marton-bod** [comment]
> Do we need to change `HiveSchemaConverter`?

Yes, good point. I'll modify the `HiveSchemaConverter` as well as the `HiveSchemaUtil` classes. Adjusting their unit tests are tricky, because the hive-metastore unit tests are run using Hive2 classes. I'll need to think how to go around that.

File: hive3/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspectorHive3.java [hunk 5]
```diff
@@ -20,50 +20,38 @@
...
   private IcebergTimestampObjectInspectorHive3() {
     super(TypeInfoFactory.timestampTypeInfo);
   }
 
-
-  abstract LocalDateTime toLocalDateTime(Object object);
+  @Override
+  public LocalDateTime convert(Object o) {
+    return o == null ? null : LocalDateTime.ofEpochSecond(
+        ((TimestampWritableV2) o).getTimestamp().toEpochSecond(), 0, ZoneOffset.UTC);
```
Comment (pvary) [inline hive3/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspectorHive3.java:46]: Do we lose nanos here?

Comment (marton-bod) [reply hive3/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergTimestampObjectInspectorHive3.java]: we did, but we shouldn't :) fixed

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveSchemaConverter.java [hunk 6]
```diff
@@ -99,6 +99,10 @@ Type convertType(TypeInfo typeInfo) {
           case INTERVAL_YEAR_MONTH:
           case INTERVAL_DAY_TIME:
           default:
+            // special case for Timestamp with Local TZ which is only available in Hive3
+            if ("TIMESTAMPLOCALTZ".equals(((PrimitiveTypeInfo) typeInfo).getPrimitiveCategory().name())) {
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveSchemaConverter.java:103]: `equalsIgnoreCase`?

**Comment by rdblue** [comment]
@pvary, please ping me when this is ready to merge.

File: hive3/src/test/java/org/apache/iceberg/mr/hive/TestHiveSchemaUtilHive3.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.hive.TestHiveSchemaUtil;
+import org.apache.iceberg.types.Types;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+
+public class TestHiveSchemaUtilHive3 extends TestHiveSchemaUtil {
+
+  @Override
+  protected List<FieldSchema> getSupportedFieldSchemas() {
```
Comment (pvary) [inline hive3/src/test/java/org/apache/iceberg/mr/hive/TestHiveSchemaUtilHive3.java:35]: Is it possible to use super and add only specific fields?

File: hive3/src/test/java/org/apache/iceberg/mr/hive/serde/objectinspector/TestIcebergTimestampObjectInspectorHive3.java [hunk 8]
```diff
@@ -51,6 +48,7 @@ public void testIcebergTimestampObjectInspector() {
     Assert.assertNull(oi.copyObject(null));
     Assert.assertNull(oi.getPrimitiveJavaObject(null));
     Assert.assertNull(oi.getPrimitiveWritableObject(null));
+    Assert.assertNull(oi.convert(null));
 
     long epochMilli = 1601471970000L;
     LocalDateTime local = LocalDateTime.ofInstant(Instant.ofEpochMilli(epochMilli), ZoneId.of("UTC"));
```
Comment (pvary) [inline hive3/src/test/java/org/apache/iceberg/mr/hive/serde/objectinspector/TestIcebergTimestampObjectInspectorHive3.java:54]: We do not check nanos. Do we want to do it?

Comment (rdblue) [reply hive3/src/test/java/org/apache/iceberg/mr/hive/serde/objectinspector/TestIcebergTimestampObjectInspectorHive3.java:54]: Yes!

Comment (marton-bod) [reply hive3/src/test/java/org/apache/iceberg/mr/hive/serde/objectinspector/TestIcebergTimestampObjectInspectorHive3.java]: Done!

File: hive3/src/test/java/org/apache/iceberg/mr/hive/TestHiveSchemaUtilHive3.java (hunk 7 above)
Comment (marton-bod) [reply hive3/src/test/java/org/apache/iceberg/mr/hive/TestHiveSchemaUtilHive3.java:35]: yes, changed it to use super

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveSchemaConverter.java (hunk 6 above)
Comment (marton-bod) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveSchemaConverter.java]: I think it should always be all uppercase, but I've changed it just to be safe.

**Comment by rdblue** [comment]
Merged. Thanks @marton-bod, and thanks for reviewing, @pvary!

**Pull request merged**